### PR TITLE
`dune install` should handle destination directories that don't exist

### PR DESCRIPTION
**Pull request by hongchangwu**
This is a follow-up to #1070.

There are still several problems with `dune-install`:

1. `dune install --prefix $PREFIX` will complain if `$PREFIX` doesn't exist.
2. `dune install --prefix $PREFIX --libdir lib` will complain even if `$PREFIX/lib` already exists (because `libdir` is a `dir` argument).

In both cases, `opam-installer` will happily install and create the directory if it doesn't exist. This PR brings `dune-install` closer to the behavior of `opam-installer`.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Looks good. We should add tests for this

**Comment by emillon** [comment]
More generally, I think that we should get rid of `dir` because it does not work when using `--root` (it only checks relatively to the current directory).

**Comment by rgrinberg** [comment]
> More generally, I think that we should get rid of dir because it does not work when using --root (it only checks relatively to the current directory).

How about we create a custom dir term that doesn't have these deficiencies?

**Comment by emillon** [comment]
I don't think it's possible with an applicative, since what is a valid "dir" would depend on the runtime value of the "root" option.

**Comment by diml** [comment]
It could work as it is not necessary to change the graph to do such a check. However, I don't think the API of cmdliner allows to define a term that depends on values computed by the applicative. In any case doing the check after parsing the command line seems acceptable. In this particular case, we want to create the directory if it doesn't exist anyway.

**Comment by hongchangwu** [comment]
Added a test for this.

File: test/blackbox-tests/test-cases/install-libdir/run.t [hunk 1]
```diff
@@ -0,0 +1,15 @@
+`dune install` should handle destination directories that don't exist
+
+  $ dune build @install
+  $ dune install --prefix /tmp --libdir lib
```
Comment (diml) [inline test/blackbox-tests/test-cases/install-libdir/run.t]: The test should delete `/tmp/lib` after this. I suggest to use a local directory instead: `--prefix install`.

Comment (diml) [reply test/blackbox-tests/test-cases/install-libdir/run.t]: Ah, that's annoying, we'll need to look into this

Comment (diml) [reply test/blackbox-tests/test-cases/install-libdir/run.t]: Could you rebase and try again with a relative prefix? I fixed the problem with `Path.mkdir_p`. The output should be stable as it should print a relative path. Writing in `/tmp` using a non random name doesn't feel right.

Comment (emillon) [reply test/blackbox-tests/test-cases/install-libdir/run.t]: Agree about fixed files. Plus some systems might not have `/tmp`, etc.

Comment (hongchangwu) [reply test/blackbox-tests/test-cases/install-libdir/run.t]: Thanks for the fix @diml. Will update the test.

**Comment by diml** [comment]
Merged, thanks for the PR!

**Pull request merged**